// @marko/compiler/dist/babel-plugin/plugins/migrate.js
"use strict";exports.__esModule = true;exports.visitor = void 0;var _babelUtils = require("@marko/compiler/babel-utils");

var t = _interopRequireWildcard(require("../../babel-types"));
var _pluginHooks = require("../util/plugin-hooks");function _interopRequireWildcard(e, t) {if ("function" == typeof WeakMap) var r = new WeakMap(),n = new WeakMap();return (_interopRequireWildcard = function (e, t) {if (!t && e && e.__esModule) return e;var o,i,f = { __proto__: null, default: e };if (null === e || "object" != typeof e && "function" != typeof e) return f;if (o = t ? n : r) {if (o.has(e)) return o.get(e);o.set(e, f);}for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]);return f;})(e, t);}

/**
 * Applies custom migrators on tags.
 */
const visitor = exports.visitor = {
  MarkoTag: {
    enter(path) {
      const migrators = getMigratorsForTag(path);
      const { node } = path;
      for (const migrator of migrators) {
        (0, _pluginHooks.enter)(migrator, path, t);
        if (path.node !== node) break; // Stop if node is replaced.
      }
    },
    exit(path) {
      const migrators = getMigratorsForTag(path);
      const { node } = path;
      for (const migrator of migrators) {
        (0, _pluginHooks.exit)(migrator, path, t);
        if (path.node !== node) break; // Stop if node is replaced.
      }
    }
  }
};

function getMigratorsForTag(path) {
  const {
    hub: { file }
  } = path;
  const { watchFiles } = file.metadata.marko;
  const tagName = path.get("name.value").node || "*";
  const MIGRATOR_CACHE = file.MIGRATOR_CACHE =
  file.MIGRATOR_CACHE || Object.create(null);

  let migrators = MIGRATOR_CACHE[tagName];

  if (!migrators) {
    migrators = MIGRATOR_CACHE[tagName] = [];
    const addMigrators = (tagDef) => {
      if (tagDef && tagDef.migrators) {
        for (const migrator of tagDef.migrators) {
          if (migrator.path) {
            watchFiles.push(migrator.path);
          }
          migrators.push(migrator.hook.default || migrator.hook);
        }
      }
    };

    addMigrators((0, _babelUtils.getTagDef)(path));

    if (tagName !== "*") {
      addMigrators((0, _babelUtils.getTagDefForTagName)(file, "*"));
    }
  }

  return migrators;
}
